### ISSUE TITLE
Possible bug in expectsJobs() method

### ISSUE THREAD

**Issue by MatasElectronics** [orchestral/testbench#152]
We've been using this package for a while now, and it's really great.
But there's one strange thing we just can't wrap our heads around.

When we can't to see if a Job was dispatched (or wasn't) we use the `$this->expectsJobs(Job::class);` method. But somehow, this method will only work if before that, we put something like `app('Illuminate\Contracts\Bus\Dispatcher');`

Is this a know bug or something in our codebase?

Looks like the Dispatcher first has to be booted or something, otherwise the mock won't get bound to the Dispatcher when you use the expectsJobs method.

**Comment by crynobone** [comment]
We don't actually handle this code. Let me check Laravel code.

**Comment by MatasElectronics** [comment]
Yes I know that method is not in your codebase :)

What Laravel does is create a mock of the job dispatcher and mock the 'dispatch' and 'dispatchNow' method. Next it will bind that mock to the interface, so that each call to `$this->dispatch()` or the `dispatch()` helper goes to the mock.

Problem almost looks like the mock is being bound to the interface, but somewhere between there and the actual call to the dispatch method, it gets rebound to the original dispatcher again.

Hope that makes sense ;)

**Comment by crynobone** [comment]
The only time we would flush the app binding is when executing `tearDown()`, right after running all after app destroy callback events. Again,will have to check Laravel code on this.

**Comment by MatasElectronics** [comment]
Okay, will wait on that. Thanks!

**Comment by crynobone** [comment]
Basically there might be a bug resolving around `Application::make()` and the way it handle `alias()` and `instance()`. 

Original issue: https://github.com/orchestral/testbench/issues/152

Basically what happen is `Illuminate\Bus\BusServiceProvider` declares `Dispatcher`, `DispatcherContract`, `QueueingDispatcherContract`. and alias `DispatcherContract` to `Dispatcher`. See https://github.com/laravel/framework/blob/5.4/src/Illuminate/Bus/BusServiceProvider.php

Now when you use `expectsJobs()` https://github.com/laravel/framework/blob/5.4/src/Illuminate/Foundation/Testing/Concerns/MocksApplicationServices.php it would actually use `$app->instance()` to bind a mocked `DispatcherContract` to the app, and register the expected jobs.

When you assert/run the test, `dispatch()` would be called and run `$app->make(DispatcherContract)->dispatch($job)` would be called. This would use `Application::make()` https://github.com/laravel/framework/blob/5.4/src/Illuminate/Foundation/Application.php#L694-L703. `getAlias()` would still return as `DispatcherContract`, however it is listed as a deferredServices and therefore `BusServiceProvider` will be register and booted by the app cause the issue faced by the user.

---

Have forwarded this to @taylorotwell and hoping to get some response from him soon.

**Comment by crynobone** [comment]
@MatasElectronics can you test the issue against latest commit, or lock the version to `"orchestra/testbench": "~3.4.3"`

**Comment by MatasElectronics** [comment]
@crynobone Will do! Will report back when I'm done.

**Comment by MatasElectronics** [comment]
That seems to have done it!
I must have had something to do with deferred service providers. Thanks for the in depth explanation.

**Comment by crynobone** [comment]
Fix has been tagged.

**Comment by MatasElectronics** [comment]
Thanks a lot!
